aten/src/ATen/native/RNN.cpp

```diff
@@ -593,25 +699,182 @@ std::tuple<Tensor, Tensor> lstm_cell(
 }
 
 Tensor gru_cell(
     const Tensor& input, const Tensor& hx,
     const Tensor& w_ih, const Tensor& w_hh, const Tensor& b_ih, const Tensor& b_hh) {
-  return GRUCell{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
+  return GRUCell<CellParams>{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
 }
 
 Tensor rnn_tanh_cell(
     const Tensor& input, const Tensor& hx,
     const Tensor& w_ih, const Tensor& w_hh, const Tensor& b_ih, const Tensor& b_hh) {
-  return SimpleCell<tanh_f>{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
+  return SimpleCell<tanh_f, CellParams>{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
 }
 
 Tensor rnn_relu_cell(
     const Tensor& input, const Tensor& hx,
     const Tensor& w_ih, const Tensor& w_hh, const Tensor& b_ih, const Tensor& b_hh) {
-  return SimpleCell<relu_f>{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
+  return SimpleCell<relu_f, CellParams>{}(input, hx, CellParams{w_ih, w_hh, b_ih, b_hh});
+}
+
+// Quantized implementations
+//
+// These implementations use FBGEMM to do the i2h and h2h linear layers with
+// an int8 quantized weight. This is advantageous in small-batch-size scenarios
+// where runtime is dominated by memory fetches of the weight matrix.
+
+std::tuple<Tensor, Tensor, Tensor> quantized_lstm(
+      const Tensor& _input, TensorList hx,
+      TensorList _params, bool has_biases,
+      int64_t num_layers, double dropout_p, bool train, bool bidirectional, bool batch_first) {
+  AT_CHECK(hx.size() == 2, "lstm expects two hidden states");
+  if (at::cudnn_is_acceptable(_input)) {
+    Tensor output, hy, cy;
+    lstm_cudnn_stub(_input.type().device_type(), output, hy, cy, _input, hx, _params, has_biases,
+            num_layers, dropout_p, train, bidirectional, batch_first);
+    return std::make_tuple(output, hy, cy);
+  }
+  check_device(_input, _params, hx);
+  auto input = batch_first ? _input.transpose(0, 1) : _input;
+  AT_CHECK(has_biases, "quantized LSTM requires biases");
+  auto params = gather_quantized_params(_params);
+  auto results = _lstm_impl<FullLayer, FullBidirectionalLayer>(
+      input, params, hx[0], hx[1], num_layers, dropout_p, train, bidirectional);
+  if (batch_first) {
+    std::get<0>(results) = std::get<0>(results).transpose(0, 1);
+  }
+  return results;
+}
+
+std::tuple<Tensor, Tensor> quantized_lstm_cell(
+    const Tensor& input,
+    TensorList hx,
+    const Tensor& w_ih,
+    const Tensor& w_hh,
+    const Tensor& b_ih,
+    const Tensor& b_hh,
+    const Tensor& packed_ih,
+    const Tensor& packed_hh,
+    const Tensor& col_offsets_ih,
+    const Tensor& col_offsets_hh,
+    const Scalar scale_ih,
+    const Scalar scale_hh,
+    const Scalar zero_point_ih,
+    const Scalar zero_point_hh) {
+  QuantizedCellParams params(
+      w_ih,
+      w_hh,
+      b_ih,
+      b_hh,
+      packed_ih,
+      packed_hh,
+      col_offsets_ih,
+      col_offsets_hh,
+      scale_ih,
+      scale_hh,
+      zero_point_ih,
+      zero_point_hh);
```

Comment:
Oh wow, can we collapse those somehow? 😮

Reply:
To be honest I think I would even be in favor of using a macro here, because this is _a lot_.